fetch-in-react-practice/client-ts/src/components/Todos.tsx

```diff
@@ -0,0 +1,162 @@
+/* eslint-disable @typescript-eslint/no-unused-vars -- Remove me */
+import React, { useEffect, useState } from 'react';
+import PageTitle from './PageTitle';
+import TodoList from './TodoList';
+import TodoForm from './TodoForm';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export default function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    async function fetchTodos() {
+      try {
+        const res = await fetch('/api/todos');
+        if (!res.ok) {
+          throw new Error(`fetch Error ${res.status}`);
+        }
+        const todosList: Todo[] = await res.json();
+        setTodos(todosList);
+      } catch (err) {
+        setError(err);
+      } finally {
+        setIsLoading(false);
+      }
+    }
+    fetchTodos();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo) {
+    try {
+      const options = {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(newTodo),
+      };
+      const res = await fetch('/api/todos', options);
+      if (!res.ok) {
+        throw new Error(`fetch Error ${res.status}`);
+      }
+      const savedTodo: Todo = await res.json();
+      setTodos((prev) => prev.concat(savedTodo));
+    } catch (err) {
+      setError(err);
+    } finally {
+      setIsLoading(false);
+    }
+  }
+
+  /* Implement toggleCompleted to toggle the completed state of a todo. Hints are at the bottom of the file. */
+  async function toggleCompleted(todoId: number) {
+    const index = todos.findIndex((current) => current.todoId === todoId);
+    const newStatus = {
+      isCompleted: !todos[index].isCompleted,
+    };
+    try {
+      const options = {
+        method: 'PATCH',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(newStatus),
+      };
+      const res = await fetch(`/api/todos/${todoId}`, options);
+      if (!res.ok) {
+        throw new Error(`fetch Error ${res.status}`);
+      }
+      const savedTodo: Todo = await res.json();
+      const newTodos = todos.map((current) =>
+        current.todoId === todoId ? savedTodo : current
+      );
+      setTodos(newTodos);
+    } catch (err) {
+      setError(err);
+    } finally {
+      setIsLoading(false);
+    }
+  }
+
+  if (isLoading) {
```

Comment:
If you experiment with the UI suggested above you might want to remove this check (or modify it).